### PR TITLE
feat(kafka): add support for regex topic

### PR DESCRIPTION
**Pull request by amitco1**

### CODE REVIEW

File: src/main/scala/com/yotpo/metorikku/input/readers/kafka/KafkaInput.scala [hunk 1]
```diff
@@ -8,24 +8,38 @@ import org.apache.kafka.clients.consumer.KafkaConsumer
...
   @transient lazy val log = org.apache.log4j.LogManager.getLogger(this.getClass)
 
+
   def read(sparkSession: SparkSession): DataFrame = {
     consumerGroup match {
       case Some(group) =>
         log.info(s"creating consumer group with id $group")
         val consumer = createKafkaConsumer(group)
-        val lagWriter = new KafkaLagWriter(consumer, topic)
+        val chosen_topic = topic.getOrElse(topicPattern.getOrElse(""))
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/input/readers/kafka/KafkaInput.scala]: You can't create this here when using topic pattern... you don't know the topics... I'm not even sure you need it in the lag writer

File: src/main/scala/com/yotpo/metorikku/input/readers/kafka/KafkaInput.scala [hunk 2]
```diff
@@ -8,24 +8,38 @@ import org.apache.kafka.clients.consumer.KafkaConsumer
...
     }
 
     val bootstrapServers = servers.mkString(",")
     val inputStream = sparkSession.readStream.format("kafka")
-      .option("kafka.bootstrap.servers", bootstrapServers)
-      .option("subscribe", topic)
+    topic match {
+      case Some(regular_topic) =>
+        inputStream
+          .option("kafka.bootstrap.servers", bootstrapServers)
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/input/readers/kafka/KafkaInput.scala]: This part has nothing to do with the topics, take it outside

File: src/main/scala/com/yotpo/metorikku/input/readers/kafka/KafkaInput.scala [hunk 3]
```diff
@@ -34,12 +48,12 @@ case class KafkaInput(name: String, servers: Seq[String], topic: String, consume
     val kafkaDataFrame = inputStream.load()
     schemaRegistryUrl match {
       case Some(url) => {
-        val schemaRegistryDeserializer = new SchemaRegistryDeserializer(url, topic, schemaSubject)
+        val chosen_topic = topic.getOrElse(topicPattern.getOrElse(""))
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/input/readers/kafka/KafkaInput.scala]: You can't send topic pattern, always send topic, even if empty

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala [hunk 4]
```diff
@@ -13,59 +13,50 @@ class InstrumentationOutputWriter(props: Map[String, String],
...
     val columns = dataFrame.schema.fields.zipWithIndex
-    val indexOfKeyCol = keyColumnProperty.flatMap(col => Option(dataFrame.schema.fieldNames.indexOf(col)))
+    val indexOfValCol = valueColumnProperty.flatMap(col => Option(dataFrame.schema.fieldNames.indexOf(col)))
     val indexOfTimeCol = timeColumnProperty.flatMap(col => Option(dataFrame.schema.fieldNames.indexOf(col)))
 
     log.info(s"Starting to write Instrumentation of data frame: $dataFrameName on metric: $metricName")
     dataFrame.foreachPartition(p => {
       val client = instrumentationFactory.create()
 
+      val actualIndexOfValCol = indexOfValCol.getOrElse(columns.length - 1)
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: Add note (use last column if valuecolumn is missing

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala [hunk 5]
```diff
@@ -13,59 +13,50 @@ class InstrumentationOutputWriter(props: Map[String, String],
...
-              if (valueOfRowAtCurrentCol != null && classOf[Number].isAssignableFrom(valueOfRowAtCurrentCol.getClass)) {
-                val longValue = valueOfRowAtCurrentCol.asInstanceOf[Number].longValue()
-                val keyColumnTags = getTagsForKeyColumn(indexOfKeyCol, row)
-                val time = getTime(indexOfTimeCol, row)
-                val tags = Map("metric" -> metricName, "dataframe" -> dataFrameName) ++ keyColumnTags
+        try {
+
+          val tags = Map("metric" -> metricName, "dataframe" -> dataFrameName) ++
+            columns.filter {
+              case (column, index) => index != actualIndexOfValCol || (!row.isNullAt(indexOfTimeCol.get) && index != indexOfValCol.get)
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: indexOfTimeCol

Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: indexOfTimeCol.isDefined && index != indexOfTimeCol.get

Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: index != actualIndexOfValCol && (!indexOfTimeCol.isDefined || index != indexOfTimeCol.get)

**Comment by lyogev** [comment]
Please add readme

**Comment by lyogev** [comment]
Also update config.yaml (add valuecolumn + remove keycolumn +change replace to update)

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala [hunk 6]
```diff
@@ -13,59 +13,52 @@ class InstrumentationOutputWriter(props: Map[String, String],
...
+    val indexOfValCol = valueColumnProperty.flatMap(col => Option(dataFrame.schema.fieldNames.indexOf(col)))
     val indexOfTimeCol = timeColumnProperty.flatMap(col => Option(dataFrame.schema.fieldNames.indexOf(col)))
 
     log.info(s"Starting to write Instrumentation of data frame: $dataFrameName on metric: $metricName")
     dataFrame.foreachPartition(p => {
       val client = instrumentationFactory.create()
 
+      // use last column if valueColumn is missing
+      val actualIndexOfValCol = indexOfValCol.getOrElse(columns.length - 1)
+      val actualIndexOfTimeCol = indexOfTimeCol.getOrElse(columns.length)
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: Are you sure that's the default? It should be empty if not....

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala [hunk 7]
```diff
@@ -13,59 +13,52 @@ class InstrumentationOutputWriter(props: Map[String, String],
...
-              if (valueOfRowAtCurrentCol != null && classOf[Number].isAssignableFrom(valueOfRowAtCurrentCol.getClass)) {
-                val longValue = valueOfRowAtCurrentCol.asInstanceOf[Number].longValue()
-                val keyColumnTags = getTagsForKeyColumn(indexOfKeyCol, row)
-                val time = getTime(indexOfTimeCol, row)
-                val tags = Map("metric" -> metricName, "dataframe" -> dataFrameName) ++ keyColumnTags
+        try {
+
+          val tags = Map("metric" -> metricName, "dataframe" -> dataFrameName) ++
+            columns.filter {
+              case (column, index) => index != actualIndexOfValCol && index != actualIndexOfTimeCol
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: Mmmm we can do something more elegant

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala (hunk 6 above)
Comment (amitco1) [reply src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: if timeColumn is missing then it doesn't have an index. and its default value is set in 'getTime' function

File: README.md [hunk 8]
```diff
@@ -247,6 +247,7 @@ Metorikku sends automatically on top of what spark is already sending the follow
 * In streaming: number of processed records in batch
 
 You can also send any information you like to the instrumentation output within a metric.
+by default the last column of the schema will be the field value. all other columns will be tags
```
Comment (lyogev) [inline README.md]: 1. please update the example influxdb metric here (remove keycolumn): https://github.com/YotpoLtd/metorikku/blob/master/examples/influxdb/movies_metric.yaml

Write the following:
Add an output like below:
```yaml
output:
- dataFrameName: df
  outputType: Instrumentation
  outputOptions:
    # Optional: define a custom time for the event (works only when writing directly influxDB)
    timeColumn: ts
    # Optional: when defined this column's value will become the metric value, by default the last column is the value column
    valueColumn: val
```
Other columns that are not value or time columns will be merged together as the name of the metric. If writing directly to influxDB these will become tags.

File: README.md [hunk 9]
```diff
@@ -306,6 +307,18 @@ Check out the built-in code steps [here](src/main/scala/com/yotpo/metorikku/code
 
 *NOTE: If you added some dependencies to your custom JAR build.sbt you have to either use [sbt-assembly](https://github.com/sbt/sbt-assembly) to add them to the JAR or you can use the ```--packages``` when running the spark-submit command* 
 
+#### Watermark
```
Comment (lyogev) [inline README.md]: Please move this section to be a sub section of kafka input (#####)

File: README.md [hunk 10]
```diff
@@ -306,6 +307,18 @@ Check out the built-in code steps [here](src/main/scala/com/yotpo/metorikku/code
...
+Metorikku supports Watermark method which helps a stream processing engine to deal with latenes.
```
Comment (lyogev) [inline README.md]: latenes=late data

File: README.md [hunk 11]
```diff
@@ -306,6 +307,18 @@ Check out the built-in code steps [here](src/main/scala/com/yotpo/metorikku/code
...
+You can use watermarking by adding a new udf step in your metric:
+```yaml
+- dataFrameName: dataframe
```
Comment (lyogev) [inline README.md]: Add comment: This will become the new watermarked dataframe name.

File: README.md [hunk 12]
```diff
@@ -306,6 +307,18 @@ Check out the built-in code steps [here](src/main/scala/com/yotpo/metorikku/code
...
+  classpath: com.yotpo.metorikku.code.steps.Watermark
+  params:
+    table: my_table
```
Comment (lyogev) [inline README.md]: Comment: Watermark this table

File: README.md [hunk 13]
```diff
@@ -306,6 +307,18 @@ Check out the built-in code steps [here](src/main/scala/com/yotpo/metorikku/code
...
+    eventTime: event
```
Comment (lyogev) [inline README.md]: Comment: The column representing the event time (needs to be a TIMESTAMP or DATE column)

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala (hunk 6 above)
Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: I don't think you need this variable at all

File: src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala (hunk 7 above)
Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/instrumentation/InstrumentationOutputWriter.scala]: ```index != actualIndexOfValCol && (!indexOfTimeCol.isDefined || index != indexOfTimeCol.get)```

**Pull request merged**